File: programmers/박소민/lv3/정수삼각형.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int solution(vector<vector<int>> triangle) {
    // 바텀업으로 활용
    for (int r = triangle.size() - 2; r >= 0; r--) {
        for (int c = 0; c <= r; c++) {
            triangle[r][c] += max(triangle[r+1][c],  triangle[r+1][c+1]);
        }
    }
    
    return triangle[0][0];
}

File: programmers/박소민/lv3/정수삼각형.py


def solution(triangle):
    LEN = len(triangle)
    arr = [[0] * (i+1) for i in range(LEN)]
    
    arr[0][0] = triangle[0][0]
    
    for r in range(1, LEN):
        for c in range(r+1):
            curr = triangle[r][c]
            if c == 0:
                arr[r][c] = curr + arr[r-1][0]
            elif c == r:
                arr[r][c] = curr + arr[r-1][c-1]
            else:
                arr[r][c] = curr + max(arr[r-1][c-1], arr[r-1][c])

    return max(arr[LEN-1])